processor.go

```diff
@@ -352,33 +345,65 @@ func (g *Processor) rebalanceLoop(ctx context.Context) (rerr error) {
 	}()
 
 	for {
+		sessionCtx, sessionCtxCancel := context.WithCancel(ctx)
+
+		go func() {
+			g.handleSessionErrors(ctx, sessionCtx, sessionCtxCancel, consumerGroup)
+		}()
+
 		err := consumerGroup.Consume(ctx, topics, g)
-		var (
-			errProc  *errProcessing
-			errSetup *errSetup
-		)
+		sessionCtxCancel()
 
-		if errors.As(err, &errProc) {
-			g.log.Debugf("error processing message (non-transient), shutting down processor: %v", err)
-			return err
-		}
-		if errors.As(err, &errSetup) {
-			g.log.Debugf("setup error (non-transient), shutting down processor: %v", err)
+		// error consuming, no way to recover so we have to kill the processor
+		if err != nil {
 			return err
 		}
 
-		if err != nil {
-			g.log.Printf("Error executing group consumer (continuing execution, rebalance after short sleep): %v", err)
-		}
+		g.log.Debugf("Consumer group returned, trying rebalance")
 
 		select {
 		case <-time.After(5 * time.Second):
-		case <-ctx.Done():
+		case <-sessionCtx.Done():
 			return nil
 		}
 	}
 }
 
+func (g *Processor) handleSessionErrors(ctx, sessionCtx context.Context, sessionCtxCancel context.CancelFunc, consumerGroup sarama.ConsumerGroup) {
+	errs := consumerGroup.Errors()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-sessionCtx.Done():
```

Comment:
Either context.Done will break the loop. Is this correct or do you need to wait for both?

Reply:
yeah, because `sessionCtx` recreated on each loop iteration and starts the error-reading go-routine again. But we need to stop this go-routine on rebalance (where there is no error), otherwise we end up with multiple of those routines. The parent context (ctx) however will not be cancelled yet, because there was not an error yet.